baseservice/iodevices/input_devices.py

```diff
@@ -0,0 +1,399 @@
+from abc import ABCMeta, abstractmethod
+from threading import Event
+from time import time, sleep
+from typing import Optional, Tuple, List, Dict, Union
+
+from baseservice.iodevices.common import Message, DeviceHeaders
+from baseservice.utils import KwargsException, StatefulListIterator
+
+
+class InputDeviceException(KwargsException):
+    """
+    a base exception class for all input device related exceptions
+    """
+    pass
+
+
+class InputTransaction(metaclass=ABCMeta):
+    """
+    this is the base class for input transaction objects.
+
+    after reading a message from the device (using 'with_transaction=True') the reader should commit/rollback
+    the transaction, to signal the device if the message is done processing or not
+    """
+
+    def __init__(self, device: 'InputDevice'):
+        """
+
+        :param device: the input device that returned that transaction
+        """
+        self._device: 'InputDevice' = device
+        self._finished: Event = Event()
```

Comment:
I think 'active' will be a more readable name

Reply:
it's the exact opposite of 'active'.
it a flag that signals that the transaction has finished. you can wait on it for the transaction to finish....
nobody cares about the transaction being 'active'

Reply:
Oh right, forgot it is an event. I would consider having the boolean property though

Reply:
call the property "active" and return not self._finished.is_set()?
why? why is 'active' better then 'done' ?

Reply:
It just sounds better in my (subjective) opinion